### DISCUSSION TITLE
MongoDBAdapter is it possible to extend the User Schema using the Mongo adaper?

### DISCUSSION THREAD

**Discussion by jberrydev** [nextauthjs/next-auth#3793] [Help]
### Question 💬

Thanks in advance.  I am trying to add phoneNumber to the user entity.  I have tried to extend the User model via TypeORM Adapter, but it doesn't seem to work with Mongo.  The options for the MongoDBAdapter do not seem to allow me to pass new schema or entity.   Is there an easy way to add a property to User so that next-auth will store this in the DB?

Version of next-auth: v4

### How to reproduce ☕️

Not a bug, just a question.

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment]
You simply return the extra data from your provider's `profile` callback and it will be saved on the user https://next-auth.js.org/adapters/models#user

**Reply by jberrydev** [reply]
Ok so this isn't handled in the MongoDBAdapter?  If I understand correctly, I need to do this for each provider?  Is it possible to show a small example of this say by adding a phoneNumber property to the user.  For more clarification, I will be the one adding the phone number and not from the OAuth provider.  I understand I can grab more data from the profile, but this is a custom field that I want to add to the User schema and update it myself.

**Reply by jberrydev** [reply]
Ok at this point I will just use the MongoDBAdapter method to update the user.  https://next-auth.js.org/tutorials/creating-a-database-adapter.

**Reply by kohm** [reply]
Did you find a solution for adding more fields to the user document?

**Comment by wkd-kapsule** [comment]
I don't know if you still need it but it is pretty easy to add custom fields to the user object created when someone signs in for the first time. First you modify the User interface used by Next-Auth (create a next-auth.d.ts file to overwrite the default one) and then add the new field to the profile() callback for each provider you are using. In this example I add a 'token' property set to a random id. It's useful when you need to expose an identifier for your user (e.g. to name a cookie) without exposing their actual id from your database. Don't forget to return the default fields as well.

``` typescript
// next-auth.d.ts
import { DefaultSession, DefaultUser } from 'next-auth';

declare module 'next-auth' {
// Adding the new field to the User interface
  interface User extends DefaultUser {
    token: string;
  }

// Here I add the user object to the session object so I can access it easily.
  interface Session extends DefaultSession {
    user: User;
  }
}
```

``` typescript
// app/api/auth/[...next-auth]/route.ts (using App router)
{...}
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID as string,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET as string,
      profile(profile: GoogleProfile) {
        return {
          // Return the default fields
          id: profile.sub,
          name: profile.name,
          email: profile.email,
          image: profile.picture,
          // Add a new one
          token: uuid(),
        };
      },
    }),
  ],
  callbacks: {
    // Here I add the id and token to the user object stored in the session object.
    // Otherwise they would be correctly written in your database but wouldn't be available in the session object.
    async session({session, user}: {session: Session; user: AdapterUser}) {
      session.user.id = user.id;
      session.user.token = user.token;
      return session;
    },
  },
{...}
```

**Reply by Yolo390** [reply]
@kapsule-studio Hi there,

it's true for Oauth providers like Google or GitHub, not for magic links with Email provider. We don't have access to a "profile" object with Email provider.

For Email provider, the only way I found so far is to write my own Adapter like it's prescribe in documentation.

> You can [create your adapter](https://authjs.dev/guides/adapters/creating-a-database-adapter) if you want to use Auth.js with a database that is not supported out of the box, or you have to change fields on any of the models.

https://next-auth.js.org/tutorials/creating-a-database-adapter

In my case, I used the official MongoDB Adapter that I have extended to add more fields.
https://github.com/nextauthjs/next-auth/tree/main/packages

**Reply by kohm** [reply]
Would you please provide an example on how to add extra fields to when you use the Email provider

**Reply by vlasaAndrei** [reply]
Augmenting the interfaces User or AdapterUser triggers type errors when trying to use the options on NextAuth and getServerSession, this is the whole stack trace: 
```
Argument of type '[{ providers: OAuthConfig<GoogleProfile>[]; callbacks: { session({ session, user }: { session: Session; user: AdapterUser; }): Promise<Session>; }; adapter: Adapter; }]' is not assignable to parameter of type 'GetServerSessionParams<GetServerSessionOptions>'.
  Type '[{ providers: OAuthConfig<GoogleProfile>[]; callbacks: { session({ session, user }: { session: Session; user: AdapterUser; }): Promise<Session>; }; adapter: Adapter; }]' is not assignable to type '[GetServerSessionOptions]'.
    Type '{ providers: OAuthConfig<GoogleProfile>[]; callbacks: { session({ session, user }: { session: Session; user: AdapterUser; }): Promise<Session>; }; adapter: Adapter; }' is not assignable to type 'GetServerSessionOptions'.
      Type '{ providers: OAuthConfig<GoogleProfile>[]; callbacks: { session({ session, user }: { session: Session; user: AdapterUser; }): Promise<Session>; }; adapter: Adapter; }' is not assignable to type 'Partial<Omit<AuthOptions, "callbacks">>'.
        The types of 'adapter.createUser' are incompatible between these types.
          Type '((user: Omit<import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/@auth/core/adapters").AdapterUser, "id">) => import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/@auth/core/types").Awaitable<import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/@auth/core/adapters").Adap...' is not assignable to type '((user: Omit<import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/next-auth/adapters").AdapterUser, "id">) => import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/next-auth/core/types").Awaitable<import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/next-auth/adapters").Ad...'.
            Type '(user: Omit<import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/@auth/core/adapters").AdapterUser, "id">) => import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/@auth/core/types").Awaitable<import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/@auth/core/adapters").Adapt...' is not assignable to type '(user: Omit<import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/next-auth/adapters").AdapterUser, "id">) => import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/next-auth/core/types").Awaitable<import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/next-auth/adapters").Ada...'.
              Type 'import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/@auth/core/types").Awaitable<import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/@auth/core/adapters").AdapterUser>' is not assignable to type 'import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/next-auth/core/types").Awaitable<import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/next-auth/adapters").AdapterUser>'.
                Type 'AdapterUser' is not assignable to type 'Awaitable<AdapterUser>'.
                  Type 'import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/@auth/core/adapters").AdapterUser' is not assignable to type 'import("c:/Users/vlasq/projects/personal/workout-buddy/node_modules/next-auth/adapters").AdapterUser'.ts(2345)
 ```
